bin/varnishd/cache/cache_vrt_vcl.c

```diff
@@ -472,8 +488,40 @@ VCL_##func##_method(struct vcl *vcl, struct worker *wrk,		\
 	CHECK_OBJ_NOTNULL(vcl->conf, VCL_CONF_MAGIC);			\
 	CHECK_OBJ_NOTNULL(wrk, WORKER_MAGIC);				\
 	vcl_call_method(wrk, req, bo, specific,				\
-	    VCL_MET_ ## upper, vcl->conf->func##_func);			\
+	    VCL_MET_ ## upper, vcl->conf->func##_func, vcl->conf->nsub);\
 	AN((1U << wrk->handling) & bitmap);				\
 }
 
 #include "tbl/vcl_returns.h"
+
+/*--------------------------------------------------------------------
+ */
+
+VCL_VOID
+VRT_call(VRT_CTX, VCL_SUB sub)
+{
+	struct vbitmap *vbm;
+
+	CHECK_OBJ_NOTNULL(ctx, VRT_CTX_MAGIC);
+	CHECK_OBJ_NOTNULL(sub, VCL_SUB_MAGIC);
+	CHECK_OBJ_NOTNULL(ctx->vcl, VCL_MAGIC);
+	assert(sub->vcl_conf == ctx->vcl->conf);
+
+	if ((sub->methods & ctx->method) == 0) {
+		VRT_fail(ctx, "Dynamic call to \"sub %s{}\" not allowed "
+		     "from this context", sub->name);
```

Comment:
I think "context" is not very helpful here.

We can stash a const char * with the name of the method we started from in VRT_CTX, so that me message could say:
    ... starting from vcl_recv{}
instead of "from this context".

Reply:
we already have the method in `(struct vrt_ctx).method`. As we need the name only for the error message, we could have a generated function which returns the name for the method bitfield, asserting that only one bit is set.
Yet the error message is not emit in isolation, VSL already contains info on the called builtin sub.
Do you still think that we should have the method again in the fail message?

Reply:
maybe rephrase as _now allowed from here_?